sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala

```diff
@@ -36,14 +36,15 @@ import org.apache.spark.util.Utils
 case class CollectLimitExec(limit: Int, child: SparkPlan) extends UnaryExecNode {
   override def output: Seq[Attribute] = child.output
   override def outputPartitioning: Partitioning = SinglePartition
-  override def executeCollect(): Array[InternalRow] = child.executeTake(limit)
-  private val serializer: Serializer = new UnsafeRowSerializer(child.output.size)
+  override def requiredChildDistribution: List[Distribution] = AllTuples :: Nil
+  override def executeCollect(): Array[InternalRow] = {
+    child.collect {
```

Comment:
Also, one other idea: if you're going to try this strategy, what do you think about putting this logic into `GlobalLimitExec` and removing the `CollectLimitExec` planning logic?